### ISSUE TITLE
Podcasts being removed from folders

### ISSUE THREAD

**Issue by reginabally** [Automattic/pocket-casts-ios#791]
Internal discussions: p1679634149115029-slack-C02A333D8LQ

### Expected behavior

Podcasts added to the folders are not removed without the user taking any action.

### Actual behavior

Podcasts were removed from folders. Two reports from users and one from @joashrajin. Some of them noticed the issue after updating the app.

User reports:
* 6063249-zd-woothemes
* 6071012-zd-woothemes

##### Both issues were reported from Pocket Casts 7.34

**Comment by reginabally** [comment]
Another report in 6180668-zd-woothemes.

App Version: 7.34
Device: iPhone15,2
OS: 16.3.1

**Comment by reginabally** [comment]
Another report in 6391970-zd-woothemes. I've turned on server loggings for this user and asked them to contact us again when the issue happens.

App Version: 7.39
Device: iPhone15,2
OS: 16.5

**Comment by joashrajin** [comment]
This happened to me while travelling. I just have one or two podcasts in my folders now and the rest are out. I’ve been on some unreliable internet at times so it’s possible that the sync wasn’t successful and that is what stuck. 

Device info:

iPhone 14 Pro Max
Pocket Casts 7.40
iOS 16.5

![IMG_5229](https://github.com/Automattic/pocket-casts-ios/assets/16253818/af439e1b-5dd6-499f-b15a-4ece502b27b7)
![IMG_5228](https://github.com/Automattic/pocket-casts-ios/assets/16253818/2e9d7cea-9f77-4dab-ba7c-fae4b6981656)

**Comment by igotdes** [comment]
+1 in #6378898-zen:

> Out of the blue about 10 podcasts seem to forget which folders they belonged to.  The folders are there, they are still in order, some are still in those folders.but about 10 of them just popped out.
>
> I checked the online app and it seemed to forget ALL the folders and a series of the podcasts I had subscribed to.

**Comment by reginabally** [comment]
> My folders are empty since yesterday
Today I sorted my feeds back to the folders but now they are empty again.

I turned on server logging for this user and we requested them to contact us when the issue happens again.

6391970-zd-woothemes

**Comment by reginabally** [comment]
> Ever since updating to iOS 16.5.1, I’ve noticed podcasts randomly popping out of folders and back in the main list. I’ll put them back and many others will have popped out the next time I open the app.

I turned on server logging for this user and we requested them to contact us when the issue happens again.

6460680-zd-woothemes

**Comment by joashrajin** [comment]
+1 in 6465685-zd-woothemes

> Most of the podcasts keep popping out of their folders and its getting extremely annoying

Enabled Server Logging

**Comment by reginabally** [comment]
+1 in 6484464-zd-woothemes

> for the past couple of weeks, the podcasts will randomly be removed from the folders and I’ll have to re-assign them.

Enabled server logging.

**Comment by azollafox** [comment]
6465685-zd-woothemes just let us know it happened again:

> If it helps, it happened again just now and a podcast episode I finished last night appeared again in the queue and was unarchived and like I had never listened to it.

**Comment by joashrajin** [comment]
@leandroalonso just sharing a possible reason for this issue. I just logged into my account on my Android device, and when the podcasts start syncing, your folders are the last thing to sync. I think that maybe if the sync doesn't complete, that causes the folders to get removed or podcasts to be removed. I could be wrong, but this issue occurred for me a few times recently while on some really unreliable internet

**Comment by leandroalonso** [comment]
>  I could be wrong, but this issue occurred for me a few times recently while on some really unreliable internet

Yes, I'm almost sure that unreliable internet or slow response from the server is contributing to that!

I'm still unable to figure out why the app is "removing" the folder UUID from the podcast. In order for the podcast to "pop out" the app needs to change the folder UUID **and** save it.

Looking at the code, if the podcast has a folder, it should never change UNLESS: the user manually removes the podcast from the folder (or the folder itself) OR the server returns an updated folder UUID. I can't find a place in the code that this would happen outside these flows but it's definitely happening 😅

**Comment by leandroalonso** [comment]
I've been trying to force this issue for the past few days without success. I'll try to summarize all my findings for future reference:

1. The issue seems to be somehow related to unreliable internet and/or slow responses from the server. Mostly all of the logs of people affected by this issue have these lines:

```
Refresh failed: with error The request timed out., status code 0
Podcast refresh failed with message: none. See previous log for more details.
```

I forced this in many of my tests but couldn't reproduce the bug at all, which takes me to the second point.

2. I believe some concurrency might be involved. Due to the slow responses some calls are being made and received in different orders and that might be affecting the podcast folder;
3. One minor thing but worth noticing: iOS — due to Protobuf — uses the `hasFolderUuid` property. We check that to assign a folder to a given podcast. If for any reason the server responds `false` to that, it might cause the folder to be "discarded". That does not seem to be the case based on my tests;
4. In theory we can "brute force" a fix for this issue: we can simply only remove a folder UUID if (and only if) the server has returned no folder for a given podcast or the action was initiated by the user (either by removing the folder or removing the podcast from the folder). That would be super hacky though and we wouldn't comprehend why the issue is happening to begin with.
5. I added additional logs that might help us track down the problem in #996.

I'm still gonna devote a few more time to look into this but I really do believe that there's a special/uncommon condition that I am missing. If just an unreliable network would cause that, the issue would be way more widespread than it seems.

**Comment by joashrajin** [comment]
+1 in 6592309-zd-a8c

> I spent a good deal of time this week sorting things manually, but when I opened the app today, all but one of my folders were gone, and my podcasts were totally unsorted again.

**Comment by erricgunawan** [comment]
> I wanted to report a bug where podcasts that were previously moved to a folder were removed from the folder.

> App Version: 7.43.1
Device: iPhone15,2
OS: 16.6

6618345-zen

**Comment by igotdes** [comment]
> I sort all of my podcasts into various folders but I’m seeing that they aren’t staying in the folders all the time. Is there a fix for the bug?

- #6643747-zen

**Comment by erricgunawan** [comment]
User from 6592309-zen (https://github.com/Automattic/pocket-casts-ios/issues/791#issuecomment-1665372190) had the issue again.

Previously, on August 5:

> I was able to do the resorting today, so I've got it back to how I want it.  

But on August 8:

> Unfortunately, this has happened again today, as my podcasts got scrambled out of the folder again.

We also asked the following questions and here are the answers:

> How many devices you have logged into your Pocket Casts account? If you can also share the names of the devices, it would help!
> - Just one device - my iPhone.
>
> Are you using WiFi or Mobile Data when using Pocket Casts?
> - Both mobile data and WiFi
>
> Did you do any iOS updates around the time this happened, or did Pocket Casts update?
> - No iOS updates that I was aware of.
>
> Any steps you followed before you noticed this?
> - I don't think I did anything in particular beforehand.

**Comment by reginabally** [comment]
6643747-zen. Issue reported in v7.43.1.

```
1. 2 devices. My iPhone which may show up as “Katelyn’s iPhone” or something random as my email is automatically hidden when I log into things using my apple account. And a laptop browser using Chrome. It’s my work laptop so that was why I didn’t download the Pocket Cast desktop app but just use Chrome for it. 
2. Both. On the laptop, I typically use WiFi. My phone is mixed but more often than not, mobile data. 
3. No iOS updates coming to mind. The folder undo has happened a couple of times now which is why I reached out to support regarding it. I transferred a lot of my data from Stitcher which meant I have a lot that I have finished over the years, but my friends come to me asking for recommendations so that's why I keep those podcasts in a "finished" folder. It's also nice because then I see if the host starts a new series when they drop a preview on that old series. But my Pocket Cast updates are automatically done on my phone. I know if I put myself in charge of remembering to update them, it'll never happen so it's set to auto update. 
4. No other abnormal steps. I've added some podcasts when I was in the browser and noticed it unsynced my folders but other times besides that, it's just been switching from mobile to browser app. I have a playlist that I just run between both places that I listen throughout the day, so I am not going into my folders to find episodes during my time on the browser normally. 
```

**Comment by yaelirub** [comment]
#6694130-zen
>yesterday, I opened up > PocketCasts and then several podcasts became orphaned out of their folders. I have not opened PocketCasts on any other > device in months. 


Logs in the ticket

**Comment by joashrajin** [comment]
6702243-zd-a8c

> Almost all of my podcasts have jumped out of their folders today. I have three folders, one is now completely empty, one has 3 podcasts left in it, and the other only has 2 left.

**Comment by igotdes** [comment]
User in #6484464-zen /#6694130-zen has shared new logs with us. They also shared the following about the devices they use Pocket Casts with and their network connection when the issue occurred:

> I primarily use Pocket Casts on my iPhone. I downloaded the desktop app for my Mac a few months ago, and I think I used it for a day or two, but haven't used it since then. I likely also used my Pocket Casts account on my previous iPhone, but it's been almost a year since I started using a newer iPhone. If you think it would help for me to remove my account from and uninstall the desktop app, that's fine by me since I've found that I prefer listening on my phone. Regarding a fast internet connection: This happened while I was standing outside of our garage at home, and our garage has its own wifi access point and I think we have a fast connection at home in general.

**Comment by igotdes** [comment]
The user in #6702243-zen reported back about the devices they use with Pocket Casts and their network connection:

> I’m logged into Pocket Casts on my iPhone, Mac, and iPad, but I only ever use it on my phone. I’ve no idea if I was connected to a fast internet when it happened, as I have no idea when it happened specifically. I noticed it while I was connected to my home WiFi which is a fast connection, but I don’t know if that’s when it happened, or if it was maybe while I was out earlier in the day.

**Comment by joashrajin** [comment]
> I created a folder with dozens of podcasts (named « Giving »), but now it’s just showing up with 4. Any tips for restoring the lost podcasts?


6735993-zd-a8c

**Comment by joashrajin** [comment]
> Hi my folders are all over the place, my arrangement is similar and all of a sudden podcasts Ive unsubscribed from are back. 

6845911-zd-a8c

Edited to add the user's response when we requested more info:

> Any specific steps you followed before you notice this happen

Nothing specific, I just opened the iOS app on my phone as I do numerous times a day and was confused as my icons were not where they have been arranged to and there were numerous podcasts there I’ve unsubscribed from

> If your internet connection was stable and fast nothing notable

I was at home all day on my wifi and noticed no change in wifi service

> How many devices you have connected to your Pocket Casts account

Three, my phone, my iPad and on my MacBook. However, I only ever use it on my phone

**Comment by thabotswana** [comment]
>A bunch of already listened to and archived episodes from a variety of podcasts will suddenly appear in my listening feed. In addition the podcasts that are stored in folders will suddenly not be in their assigned folder. I am running a beta of iOS 17 but like I said this is a recent development and it has happened 3 times now so I wanted to let you know about it.

7073411-zd-a8c

**Comment by thabotswana** [comment]
>After a bit of time, my "new" podcasts started showing up, but weren't organized properly in their folders. Older podcasts remained. I ended up cleaning it up myself.
I believe it has kept my listening record up-to-date, so the main thing was the temporary removal of my current podcasts, addition of old removed ones, and the messing up of my folder structure.

7075399-zd-a8c

Editing to add more details:

> I have it on both my personal phone and work phone, but don’t use it as much on my work phone.
I had recently removed the app from my work phone - I don’t remember opening it up directly, but that would be the most likely explanation here.
Apart from that I was using it normally, and have good stable internet.

**Comment by mdrockwell** [comment]
> Hello! A few times in the past few weeks, I’ve noticed that some of my podcasts get removed from their folders and returned to the main list. I haven’t seen much discussion of this except for a Reddit thread (https://www.reddit.com/r/pocketcasts/comments/14h8ab9/podcasts_randomly_being_taken_out_of_folders/) where someone suggested reaching out to support, so here I am! I’m using iOS app version 7.48.0.4. If there’s any data or information I can provide, please let me know!

7128984-zd-a8c

**Comment by erricgunawan** [comment]
> I use Pocketcasts on my iPhone OS 17.0.3. I created a few folders and moved all of my podcasts into the folders, but once or twice a week, randomly most of the podcasts are removed from the folders

7152093-zen

**Comment by erricgunawan** [comment]
> For the second time in three days I accessed my folder to find that nearly all of my podcasts were randomly removed from their folders and I had to manually put them back in their folders.  

7152969-zen

**Comment by geekygecko** [comment]
For the users that this is repeatable for is it possible to ask them for their sync email address and for permission to turn on server logging? If they could email us when it happens again it might help us track down what is causing the issue.

**Comment by azollafox** [comment]
> My folders have suddenly partially emptied with the vast majority of podcasts being removed from folders. Interestingly, not all have been removed from folders. This seemed to have occurred when I logged into the web app on my computer, but that could just be coincidence. 

7204614-zen.  I'm asking them for permission to turn on server logging as suggested above

/Edit: server logging has been turned on for this user's account

**Comment by igotdes** [comment]
> I have created folders for my podcasts and organized my podcasts into the folder I want them to be in. The last few times I’ve opened the app, all or most of my podcasts have been removed from their folders and are just in the main podcast list. I have over 100 podcasts on my list so it’s very frustrating and time consuming to have to reorganize them every time.

- #7231237-zen

I've asked them for permission to turn on server logging for their account.

/Edit: server logging has been turned on for this user's account

**Comment by erricgunawan** [comment]
> Folders on iOS app are ..forgetting they exist.
>
> Some folder now are empty and the podcasts are now just in the main grid, it happens randomly.
>
> And if I don’t keep up with it. Folders keep losing podcasts to the main grid
>
> If I re-add them, sometimes it stays. Sometimes it doesn’t. Sometimes they go back in folders but new podcasts jump out

7261876-zen. Asked permission to turn on server logging.

**Comment by cookiesgalore** [comment]
I've been seeing this issue happen every few weeks for several months now. It's random in terms of when it happens, and how many podcasts are removed from folders. It honestly seems like there's a race condition. With that in mind, I took a quick look through the codebase, and I'm wondering if the race condition might happen when a background refresh for local changes and a full sync occur in a similar timeframe? 

It looks like the `processServerHomeGrid` function marks all the podcasts as unsynced and then discards the local folder settings in favor of the server folder settings. If, at the same time, the `performBackgroundRefresh` function calls `syncDownloadTask` which calls `incrementalSyncRequest` you're going to have unsynced episodes without local folder info. Then when `createIncrementalSyncData` is called, and in turn calls `changedPodcasts` your data records are going to be garbage and the `podcastRecord.folderUuid.value` is going to be set to `DataConstants.homeGridFolderUuid`.

But that's just what it looks like to me. You guys know your codebase better than I do, so maybe there's something I didn't see that would guard against this.

**Comment by joashrajin** [comment]
> Did it over night from yesterday to today and it really did it I lost all my folders and it’s so bad. 

7261876-zd-a8c

**Comment by leandroalonso** [comment]
@cookiesgalore thank you so much for your insights. I'll take a look at this. However, `processServerHomeGrid` should be called only after the initial login — and based on the logs we receive this function is not called. :(

Since you also mentioned you're seeing it frequently, would you mind contacting us through support so we can have your email and letting us enable server logs for your account?

**Comment by erricgunawan** [comment]
7309915-zen - follow up from https://github.com/Automattic/pocket-casts-ios/issues/791#issuecomment-1689243202

> Today at 5:22am, I opened PocketCasts and navigated to one of the folders to select a podcast. When I navigated into the folder, the number of podcasts shifted from 4 to 2. I navigated back to the main Podcasts screen from the folders, and noticed that at least 1/3 of the podcasts that I listen to were orphaned from their folders. I’m attached a screenshot; normally my Podcasts screen is all folders rather than individual podcasts. 

> App Version: 7.40.1
Device: iPhone14,7
OS: 16.5.1

Asked permission to turn on server logging.

**Comment by leandroalonso** [comment]
@cookiesgalore question: do you use Pocket Casts on the Apple watch?

**Comment by CameronMilliken** [comment]
@leandroalonso Here is their response:

> I see the Pocket Casts app on my Apple Watch and I'll see the watch controls on the screen when I'm listening to podcasts, but I rarely (if ever) interact with the Pocket Casts app on my watch.

**Comment by leandroalonso** [comment]
For reference:

We suspect that Apple Watch is the culprit of this issue. I'll investigate that in the next few days.

**Comment by cookiesgalore** [comment]
@leandroalonso Yeah, I have the app installed on my watch as well as my phone, but primarily use my phone in terms of interacting with the app.

**Comment by igotdes** [comment]
@leandroalonso Would uninstalling the Watch app be a temporary workaround for users affected by this issue?

**Comment by rviljoen** [comment]
> @leandroalonso Would uninstalling the Watch app be a temporary workaround for users affected by this issue?

@igotdes Yes, if the hypothesis is correct (that the issue lies in the Watch app code) then uninstalling the watch app would be a workaround. If the issue still happens, then we'll know it's not the watch app though, so this would be a good test, especially for users that do not use the watch app beyond basic playback control. If the Pocket Casts watch app is uninstalled the detail watch now playing controls will be presented by WatchOS.

**Comment by igotdes** [comment]
Thanks, @rviljoen!

I do have a new report in #7320731-zen, so I'll ask them about the Apple Watch.

**Comment by joashrajin** [comment]
> Another issue on my side, but when I was signing into the PocketCasts app on a new device (iPad) a bunch of my podcasts were taken out of their saved folders and rearranged on my Home Screen - and this issue seems to have now been synced across all my other devices.

8207267-zd-a8c

cc: @leandroalonso

**Comment by leandroalonso** [comment]
@joashrajin

Update: it seems that the issue is because they logged back in. [The app then marked all their podcasts as unsynced](https://github.com/Automattic/pocket-casts-ios/blob/trunk/podcasts/SyncSigninViewController.swift#L334) which caused old data to be synced.

--

I looked into this user's log and this is not the old bug coming back, but more of a chain of events related to this issue happening in the past + another issue.

Based on their logs, it seems that they don't use the iPad app that frequently. What happened is that it seems at some point they were hit with the old watch-folder bug.

Then last Friday they opened the app and, for some reason, the app synced all their old podcast data (with missing folders) as if it were new:

```
2024-05-17 18:17:14 [Folders] SyncTask: Number of changed podcasts: 33
```

This should not happen and it's a part of another bug in which old data is synced as if it was new. I'll look into that.
